deeprank2/query.py

```diff
@@ -416,114 +419,92 @@ def __iter__(self) -> Iterator[DeepRankQuery]:
+                used to generate features (given as string or as an imported module).
+                Each module must implement the :py:func:`add_features` function, and all feature modules must exist inside `deeprank2.features` folder.
                 If set to 'all', all available modules in `deeprank2.features` are used to generate the features.
-                Defaults to only the basic feature modules `deeprank2.features.components` and `deeprank2.features.contact`.
-            cpu_count (int | None, optional): How many processes to be run simultaneously. Defaults to None, which takes all available cpu cores.
-            combine_output (bool, optional): For combining the HDF5 files generated by the processes. Defaults to True.
+                Defaults to the two primary feature modules `deeprank2.features.components` and `deeprank2.features.contact`.
+            cpu_count (int | None, optional): The number of processes to be run in parallel (i.e. number of CPUs used), capped by
+                the number of CPUs available to the system.
+                Defaults to None, which takes all available cpu cores.
+            combine_output (bool, optional):
+                if `True` (default): all processes are combined into a single HDF5 file.
+                If `False`: separate HDF5 files are created for each process (i.e. for each CPU used).
             grid_settings (:class:`GridSettings` | None, optional): If valid together with `grid_map_method`, the grid data will be stored as well.
                 Defaults to None.
             grid_map_method (:class:`MapMethod` | None, optional): If valid together with `grid_settings`, the grid data will be stored as well.
                 Defaults to None.
-            grid_augmentation_count (int, optional): Number of grid data augmentations. May not be negative be zero or a positive number.
+            grid_augmentation_count (int, optional): Number of grid data augmentations (must be >= 0).
                 Defaults to 0.
 
         Returns:
             list[str]: The list of paths of the generated HDF5 files.
         """
 
         # set defaults
-        if prefix is None:
-            prefix = "processed-queries"
-        elif prefix.endswith('.hdf5'):
-            prefix = prefix[:-5]
-        if cpu_count is None:
-            cpu_count = os.cpu_count()  # returns the number of CPUs in the system
-        else:
-            cpu_count_system = os.cpu_count()
-            if cpu_count > cpu_count_system:
-                _log.warning(f'\nTried to set {cpu_count} CPUs, but only {cpu_count_system} are present in the system.')
-                cpu_count = cpu_count_system
-        self.cpu_count = cpu_count
-        _log.info(f'\nNumber of CPUs for processing the queries set to: {self.cpu_count}.')
+        self._prefix = "processed-queries" if not prefix else re.sub('.hdf5$', '', prefix)  # scrape extension if present
```

Comment:
This is regarding `prefix`, right? Not `cpu_count`?
I don't understand what you mean with this. (It might be easier to explain in person).

In any case, I made it an attribute, because it is called by an internal method of `QueryCollection` and thought it would be easier not to have pass a variable that is a fixed variable initialized in the class. But can be changed back if that is preferred.

Reply:
Indeed let's look at it together

Reply:
This was about `_prefix`, `_cpu_count`, `_grid_settings`, and the other internal attributes that are defined for the first time in the `process()` method. Before this PR, they were just local variables in the method, now they are internal attributes. It's not a good practice to define attributes for the first time in some method of a class that is not the init (see my comment above). I see why you did it, i.e., for readability reasons for the developers; now the `_process_one_query` method is much clearer indeed. I am not sure in this case which is the best thing to do. Maybe we can ask Lourens and/or Cunliang?

Reply:
Wouldn't the right thing be to add them all to the init of `QueryCollection`, and initialize them at `None`? And then make a better docstring explaining what they are and that they will be set properly in process? Then we both have them readable as attributes of `QueryCollection`, but also have a clean `_process_one_query` method

Reply:
This was indeed what I suggested with my very first comment, fine for me :)